File: drupal/ingest.py
#!/usr/bin/env python3
"""
Drupal module registry ingester.
Uses drupal.org REST API to list modules; downloads tarballs from ftp.drupal.org.

Usage:
    python -m drupal.ingest --max-modules 200 --core-compat 10.x --output-dir output/
    python -m drupal.ingest --core-compat 9.x --output-dir output/
"""
import argparse
import logging
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from common.state import DownloadState
from common.http import RateLimitedSession

ECOSYSTEM = "drupal"
DRUPAL_API = "https://www.drupal.org/api-d7/node.json"
PAGE_SIZE = 100

logger = logging.getLogger(__name__)


def list_modules(session: RateLimitedSession, page: int):
    """Fetch one page of full Drupal modules. Returns (modules_list, next_url_or_None)."""
    params = {
        "type": "project_module",
        "field_project_type": "full",
        "limit": PAGE_SIZE,
        "page": page,
        "sort": "changed",
        "direction": "DESC",
    }
    resp = session.get(DRUPAL_API, params=params)
    data = resp.json()
    return data.get("list", []), data.get("next")


def get_latest_release_url(session: RateLimitedSession, nid: int, core_compat: str,
                           slug: str = None):
    """Return (version, download_url) for the latest stable release matching core_compat.

    Note on Drupal versioning: D8/D9/D10 compatible modules all use the "8.x-" version
    prefix (legacy convention). D7 modules use "7.x-". Newer modules use semantic
    versioning (e.g. 1.5.0). The real API returns field_collection_item references for
    file entries, not embedded URLs, so we construct the URL from slug+version using the
    well-known ftp.drupal.org pattern.
    """
    # D7 modules: require "7.x-" prefix. D8/9/10: accept "8.x-" or semantic versioning.
    is_d7 = core_compat.startswith("7")
    params = {
        "type": "project_release",
        "field_release_project": nid,
        "limit": 10,
        "sort": "changed",
        "direction": "DESC",
    }
    resp = session.get(DRUPAL_API, params=params)
    for release in resp.json().get("list", []):
        version = release.get("field_release_version", "")
        if not version or "-dev" in version:
            continue
        if is_d7:
            if not version.startswith("7.x-"):
                continue
        else:
            # Accept 8.x- prefix (D8/D9/D10) or semantic versioning; reject D7-only
            if version.startswith("7.x-"):
                continue

        # Try embedded URL first (used in unit tests with mocked data)
        for f in release.get("field_release_files", []):
            url = f.get("file", {}).get("url", "")
            if url.endswith(".tar.gz") or url.endswith(".zip"):
                return version, url

        # Real API: field_release_files are field_collection_item refs with no embedded URL.
        # Construct the URL directly from the known ftp.drupal.org pattern.
        if slug:
            return version, f"https://ftp.drupal.org/files/projects/{slug}-{version}.tar.gz"

    return None, None


def download_module(session: RateLimitedSession, state: DownloadState,
                    module: dict, core_compat: str, output_dir: str) -> bool:
    slug = module.get("field_project_machine_name")
    nid = module.get("nid")
    if not slug or not nid:
        return False

    version, url = get_latest_release_url(session, nid, core_compat, slug=slug)
    if not url:
        logger.debug("No %s release for %s", core_compat, slug)
        return False

    if state.is_downloaded(ECOSYSTEM, slug, version):
        logger.debug("Skip %s %s", slug, version)
        return False

    ext = ".tar.gz" if url.endswith(".tar.gz") else ".zip"
    dest = str(Path(output_dir) / ECOSYSTEM / f"{slug}.{version}{ext}")
    try:
        file_hash = session.download(url, dest)
        state.record(ECOSYSTEM, slug, version, dest, file_hash)
        logger.info("Downloaded %s %s -> %s", slug, version, dest)
        return True
    except Exception as exc:
        logger.warning("Failed %s: %s", slug, exc)
        return False


def main():
    parser = argparse.ArgumentParser(description="Ingest Drupal modules")
    parser.add_argument("--max-modules", type=int, default=0,
                        help="Stop after N downloads (0 = unlimited)")
    parser.add_argument("--output-dir", default="output")
    parser.add_argument("--db", default="data/downloads.db")
    parser.add_argument("--core-compat", default="10.x",
                        choices=["7.x", "8.x", "9.x", "10.x"])
    parser.add_argument("--requests-per-second", type=float, default=1.0)
    parser.add_argument("-v", "--verbose", action="store_true")
    args = parser.parse_args()

    logging.basicConfig(
        level=logging.DEBUG if args.verbose else logging.INFO,
        format="%(asctime)s %(levelname)s %(message)s",
    )

    session = RateLimitedSession(args.requests_per_second)
    state = DownloadState(args.db)

    downloaded = 0
    page = 0
    has_more = True

    while has_more:
        try:
            modules, next_url = list_modules(session, page)
        except Exception as exc:
            logger.error("Page %d failed: %s", page, exc)
            break

        if not modules:
            break

        for module in modules:
            if args.max_modules and downloaded >= args.max_modules:
                logger.info("Reached --max-modules=%d, stopping.", args.max_modules)
                has_more = False
                break
            if download_module(session, state, module, args.core_compat, args.output_dir):
                downloaded += 1

        if has_more:
            has_more = bool(next_url)
            page += 1

    logger.info("Done. %d downloaded this run. DB totals: %s", downloaded, state.stats())


if __name__ == "__main__":
    main()

File: ts/src/drupal/ingest.ts
/**
 * Drupal module registry ingester.
 * Downloads tarballs from ftp.drupal.org and captures module metadata.
 *
 * Usage:
 *   npx tsx src/drupal/ingest.ts --max-modules 200 --core-compat 10.x
 */
import { join } from "path";
import { program } from "commander";
import { RateLimitedClient } from "../common/http.js";
import { DownloadState } from "../common/state.js";
import type { PackageMetadata } from "../common/types.js";

export const ECOSYSTEM = "drupal";
const DRUPAL_API = "https://www.drupal.org/api-d7/node.json";
const PAGE_SIZE = 100;

interface DrupalModule {
  nid: string | number;
  title: string;
  field_project_machine_name: string;
  body?: { value?: string; summary?: string };
  created?: string | number;
  changed?: string | number;
  field_project_type?: string;
  field_project_has_releases?: boolean;
  field_project_homepage?: { url?: string };
  field_project_license?: { title?: string };
}

interface DrupalRelease {
  field_release_version: string;
  changed?: string | number;
}

interface DrupalListResponse<T> {
  list?: T[];
  next?: string;
}

export async function listModules(
  client: RateLimitedClient,
  page: number
): Promise<{ modules: DrupalModule[]; hasMore: boolean }> {
  const params = new URLSearchParams({
    type: "project_module",
    field_project_type: "full",
    limit: String(PAGE_SIZE),
    page: String(page),
    sort: "changed",
    direction: "DESC",
  });
  const data = await client.getJson<DrupalListResponse<DrupalModule>>(
    `${DRUPAL_API}?${params}`
  );
  return {
    modules: data.list ?? [],
    hasMore: Boolean(data.next),
  };
}

/**
 * Return the latest stable release for a module.
 * D8/D9/D10 compatible modules use the "8.x-" version prefix.
 * D7 modules use "7.x-". Newer modules use semantic versioning.
 * URL is constructed directly from the known ftp.drupal.org pattern.
 */
export async function getLatestRelease(
  client: RateLimitedClient,
  nid: string | number,
  coreCompat: string,
  slug: string
): Promise<{ version: string; url: string } | null> {
  const isD7 = coreCompat.startsWith("7");
  const params = new URLSearchParams({
    type: "project_release",
    field_release_project: String(nid),
    limit: "10",
    sort: "changed",
    direction: "DESC",
  });
  const data = await client.getJson<DrupalListResponse<DrupalRelease>>(
    `${DRUPAL_API}?${params}`
  );

  for (const release of data.list ?? []) {
    const version = release.field_release_version ?? "";
    if (!version || version.includes("-dev")) continue;
    if (isD7) {
      if (!version.startsWith("7.x-")) continue;
    } else {
      // D8/D9/D10: accept 8.x- prefix or semantic versioning; skip D7-only
      if (version.startsWith("7.x-")) continue;
    }
    return {
      version,
      url: `https://ftp.drupal.org/files/projects/${slug}-${version}.tar.gz`,
    };
  }
  return null;
}

export function toMetadata(mod: DrupalModule): PackageMetadata {
  const slug = mod.field_project_machine_name;
  const createdTs = mod.created ? Number(mod.created) * 1000 : undefined;
  const changedTs = mod.changed ? Number(mod.changed) * 1000 : undefined;

  return {
    ecosystem: ECOSYSTEM,
    slug,
    name: mod.title,
    description: mod.body?.summary ?? mod.body?.value?.slice(0, 500),
    homepageUrl: mod.field_project_homepage?.url,
    license: mod.field_project_license?.title,
    registryAddedDate: createdTs ? new Date(createdTs).toISOString() : undefined,
    lastUpdatedDate: changedTs ? new Date(changedTs).toISOString() : undefined,
    rawMetadata: JSON.stringify({
      nid: mod.nid,
      project_type: mod.field_project_type,
      has_releases: mod.field_project_has_releases,
    }),
  };
}

export async function downloadModule(
  client: RateLimitedClient,
  state: DownloadState,
  mod: DrupalModule,
  coreCompat: string,
  outputDir: string
): Promise<boolean> {
  const slug = mod.field_project_machine_name;
  const nid = mod.nid;
  if (!slug || !nid) return false;

  const release = await getLatestRelease(client, nid, coreCompat, slug);
  if (!release) return false;

  const { version, url } = release;
  if (state.isDownloaded(ECOSYSTEM, slug, version)) return false;

  const destPath = join(outputDir, ECOSYSTEM, `${slug}.${version}.tar.gz`);
  try {
    const fileHash = await client.download(url, destPath);
    state.record(ECOSYSTEM, slug, version, destPath, fileHash);
    return true;
  } catch (err) {
    console.warn(`[drupal] Failed ${slug}: ${(err as Error).message}`);
    return false;
  }
}

async function main() {
  program
    .option("--max-modules <n>", "stop after N downloads (0 = unlimited)", "0")
    .option("--core-compat <v>", "Drupal core version", "10.x")
    .option("--output-dir <dir>", "output directory", "output")
    .option("--db <path>", "sqlite db path", "data/downloads.db")
    .option("--requests-per-second <n>", "rate limit", "1")
    .option("-v, --verbose", "verbose logging")
    .parse();

  const opts = program.opts();
  const maxModules = parseInt(opts.maxModules);

  const client = new RateLimitedClient(parseFloat(opts.requestsPerSecond));
  const state = new DownloadState(opts.db);
  const seenSlugs = new Set<string>();

  let downloaded = 0;
  let page = 0;
  let hasMore = true;

  while (hasMore) {
    let modules: DrupalModule[];
    let nextHasMore: boolean;
    try {
      ({ modules, hasMore: nextHasMore } = await listModules(client, page));
    } catch (err) {
      console.error(`[drupal] Page ${page} failed: ${(err as Error).message}`);
      break;
    }

    if (modules.length === 0) break;

    for (const mod of modules) {
      const slug = mod.field_project_machine_name;
      if (slug) {
        seenSlugs.add(slug);
        state.upsertMetadata(toMetadata(mod));
      }

      if (maxModules > 0 && downloaded >= maxModules) {
        console.log(`[drupal] Reached --max-modules=${maxModules}, stopping.`);
        hasMore = false;
        break;
      }
      if (await downloadModule(client, state, mod, opts.coreCompat, opts.outputDir)) {
        downloaded++;
        console.log(`[drupal] Downloaded ${slug}`);
      }
    }

    if (hasMore) {
      hasMore = nextHasMore;
      page++;
    }
  }

  const removed = state.markRemovedIfAbsent(ECOSYSTEM, seenSlugs);
  if (removed > 0) console.log(`[drupal] Marked ${removed} modules as removed.`);
  console.log(`[drupal] Done. ${downloaded} downloaded. DB totals:`, state.stats());
  console.log(`[drupal] Metadata tracked:`, state.metadataStats());
  state.close();
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
